pkg/apis/pipeline/v1beta1/when_types.go

```diff
@@ -107,9 +107,9 @@ func (wes WhenExpressions) HaveVariables() bool {
 // ReplaceWhenExpressionsVariables interpolates variables, such as Parameters and Results, in
 // the Input and Values. Returns a new instance of WhenExpressions.
```

Comment:
`Returns a new instance of WhenExpressions` is not true anymore, is it ?

Reply:
I think we should move toward LESS mutation and side effects in these functions vs more

Reply:
right, I will update the comment, there is no need to create new instances here. The goal is to get the params/task results resolved which can be effectively done in place.

Reply:
probably better we don't act on this in this PR but I'm advocating for the opposite.

while I don't think we should try to program in a purely functional way, I think limiting side effects and having more "pure" functions (https://dev.to/jamesrweb/principles-of-functional-programming-4b7c) leads to code that is easier to reason about and maintain over time

"functional core, imperative shell" is another approach i like: basically the idea is to limit mutations and other side effects in the code so you know exactly where they are happening.

in this specific example, its actually not clear to me why making "ReplaceWhenExpressionsVariables" mutate state instead of returning a new object is fixing anything - in the places where this function is being called (https://github.com/tektoncd/pipeline/blob/c12b84311e6fa5b7bc9c3d005c705c9944909be3/pkg/reconciler/pipelinerun/resources/apply.go#L81 https://github.com/tektoncd/pipeline/blob/c12b84311e6fa5b7bc9c3d005c705c9944909be3/pkg/reconciler/pipelinerun/resources/apply.go#L100) the returned value is being used - the core problem is probably that it isn't being used consistently, e.g. in ApplyReplacements `p` is being returned although `p` is not being directly mutated:

https://github.com/tektoncd/pipeline/blob/c12b84311e6fa5b7bc9c3d005c705c9944909be3/pkg/reconciler/pipelinerun/resources/apply.go#L87-L104

I think the inconsistent behavior we are seeing is because of the append semantics (`tasks := append(p.Tasks, p.Finally...)`) that we've debated previously - but i think the answer is not MORE mutation, i think it is relying on it LESS

anyway long story short, although we dont NEED to create new instances, I think this is a cleaner approach in the long run, but let's follow up on that in subsequent PRs